cgi/cgiutils.c

```diff
@@ -1494,6 +1494,34 @@ char *escape_string(const char *input) {
 	}
 
 
+/* url decode string for UI usage */
+char *url_decode_string(const char *input) {
+	int		len;
+	const char	*istp;
+	char		*ostp;
+	char		*output;
+
+	if(NULL == input)
+		return "";
+	len = (int)strlen(input);
+	if((ostp = output = (char *)malloc(len + 1)) == NULL)
+		return "";
+	istp = input;
+	do {
+		if((*istp == '%') &&
+		   (((*(istp + 1) >= '0') && (*(istp + 1) <= '9')) || ((*(istp + 1) >= 'A') && (*(istp + 1) <= 'F'))) &&
+		   (((*(istp + 2) >= '0') && (*(istp + 2) <= '9')) || ((*(istp + 2) >= 'A') && (*(istp + 2) <= 'F')))) {
+			istp++;
+			*ostp++ = (char)((-(int)((*istp >= 'A') ? '7' : '0') + (int)*istp++) * 16 +
+					 -(int)((*istp >= 'A') ? '7' : '0') + (int)*istp++);
```

Comment:
These 3 lines don't look right to me.  For one thing, the calculation of the ostp character is assuming the compiler will do pretty much left-to-right evaluation of the long expression.  But with embedded ++ operators and the ability of the compiler to re-order subexpression evaluation, I doubt we can stuff enough parentheses into that calculation to fully control the evaluation order.  So I would suggest using istp[1] and istp[2] in the calculation of the ostp character, and istp+=3 after that.

Also, how would something as simple as %12 be decoded?  That would take the ascii character '1' and multiply it by 16, then add the ascii character '2' to that.  Which is probably not what you want -- you need to subtract the ascii value of '0' from each of those characters.  Similarly, '7' and '0' should probably be just 7 and 0.

Reply:
Hmmm, maybe I got that wrong about subtracting '0' -- it would be implicit in the '7' and '0' you have there already.

Reply:
To my eye, re-odering the clauses would make this easier to understand:
```
*ostp++ = (char)(
    ((int)istp[1] - (int)((istp[1] >= 'A') ? '7' : '0')) * 16
    +
    ((int)istp[2] - (int)((istp[2] >= 'A') ? '7' : '0')) 
);
istp+=3;
```

Reply:
In the presence of the '++' operator, it will be left-to-right... it has no choice.

The logic for character offsets is correct.

Reply:
Revised.